Python/Product/VSInterpreters/PipPackageManager.cs

```diff
@@ -0,0 +1,651 @@
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Text.RegularExpressions;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.PythonTools.Analysis;
+using Microsoft.PythonTools.Infrastructure;
+
+namespace Microsoft.PythonTools.Interpreter {
+    class PipPackageManager : IPackageManager, IDisposable {
+        private IPythonInterpreterFactory _factory;
+        private string _indexUrl;
+        private PipPackageCache _cache;
+        private readonly Timer _refreshIsCurrentTrigger;
+        private readonly List<FileSystemWatcher> _libWatchers;
+        private readonly List<PackageSpec> _packages;
+        private CancellationTokenSource _currentRefresh;
+        private bool _isReady, _everCached;
+
+        internal readonly SemaphoreSlim _working = new SemaphoreSlim(1);
+
+        private int _suppressCount;
+        private bool _isDisposed;
+
+        // Prevent too many concurrent executions to avoid exhausting disk IO
+        private static readonly SemaphoreSlim _concurrencyLock = new SemaphoreSlim(4);
+
+        private static readonly KeyValuePair<string, string>[] UnbufferedEnv = new[] {
+            new KeyValuePair<string, string>("PYTHONUNBUFFERED", "1")
+        };
+
+        private static readonly Regex PackageNameRegex = new Regex(
+            "^(?!__pycache__)(?<name>[a-z0-9_]+)(-.+)?",
+            RegexOptions.Compiled | RegexOptions.CultureInvariant | RegexOptions.IgnoreCase);
+
+
+        public PipPackageManager(bool allowFileSystemWatchers = true) {
+            _packages = new List<PackageSpec>();
+
+            if (allowFileSystemWatchers) {
+                _libWatchers = new List<FileSystemWatcher>();
+                _refreshIsCurrentTrigger = new Timer(RefreshIsCurrentTimer_Elapsed);
+            }
+        }
+
+        public void SetInterpreterFactory(IPythonInterpreterFactory factory) {
+            if (factory == null) {
+                throw new ArgumentNullException("factory");
+            }
+            if (!File.Exists(factory.Configuration?.InterpreterPath)) {
+                throw new NotSupportedException();
+            }
+
+            _factory = factory;
+
+            // TODO: Get index from factory
+            //_indexUrl = "https://pypi.python.org/pypi/";
+            _indexUrl = null;
```

Comment:
>_indexUrl [](start = 12, length = 9)

Does this need to happen?  Can we ifdef out this and the other code which uses this because it's not implemented?